apps/web/public/sw.js

```diff
@@ -0,0 +1,384 @@
+  if (!DB) {
+    await openDB();
+  }
+  const tx = DB.transaction(EMAIL_STORE.name, "readwrite");
+  const store = tx.objectStore(EMAIL_STORE.name);
+
+  for (const mail of mails.emails) {
+    const req = store.put(mail);
+    req.onsuccess = console.log("added mail");
+    req.onerror = (e) =>
+      console.log("error occurred while adding mails to index", e);
+  }
+};
+
+async function fetchAndUpdateMails(request, decreasing) {
+  const resp = await fetch(request);
+  const restMails = await resp.json();
+  await saveMails(restMails);
+  const allMails = await loadLocalMail(decreasing);
+  messageClients(MESSAGES.MAILS_UPDATED, allMails);
+}
+
+/* Analytics/ Stats Part */
+
+function getStatsByPeriod(options) {
+  return new Promise(async (resolve, reject) => {
+    const { period, fromDate, toDate } = options;
+    // TODO: Figure out a way to use import statment in service worker without causing error, so you can use lodash in service worker ->Uncaught SyntaxError: Cannot use import statement outside a module (at service-worker.js:1:1)
+
+    const keyRange = IDBKeyRange.bound(+fromDate, +toDate, true, true);
+
+    const localData = await getLocalMail(false, keyRange);
+
+    resolve(getAllMailsClusterdbyPeriod(+fromDate, +toDate, period, localData));
+  });
+}
+
+// all inclusive
+const numberOfClusters = (start, end, size) => {
+  const window = (end - start + 1) / size;
+  return Math.floor(window) + (window - Math.floor(window) > 0 ? 1 : 0);
+};
+
+const getClusteredArray = (start, end, size) => {
+  const nClusters = numberOfClusters(start, end, size);
+  const clusters = [];
+  for (let i = 0; i < nClusters; i++) {
+    clusters.push({
+      startOfPeriod: new Date(
+        i < nClusters - 1 ? start + i * size : end,
+      ).toLocaleDateString("en-US", {
+        month: "short",
+        day: "2-digit",
+        year: "numeric",
+      }),
+      All: 0,
+      Sent: 0,
+      Read: 0,
+      Unread: 0,
+      Archived: 0,
+      Unarchived: 0,
+    });
+  }
+  return clusters;
+};
+
+function getAllMailsClusterdbyPeriod(fromDate, toDate, period, data) {
+  if (!data || !data.length) return;
+
+  const clusters = getClusteredArray(
+    fromDate,
+    toDate ?? fromDate,
+    TIME_CONSTANT[period],
+  );
+  // const
+  for (item of data) {
+    const clusterIndex = Math.floor(
+      (item.timestamp - fromDate) / TIME_CONSTANT[period],
+    );
+    const cluster = clusters[clusterIndex];
+    // All ++
+    cluster.All++;
+
+    // Read/Unread
+    item.read ? cluster.Read++ : cluster.Unread++;
+
+    // Sent
+    item.sent || cluster.Sent++;
+
+    // Inbox/Archived
+    item.inbox ? cluster.Unarchived++ : cluster.Archived++;
+  }
+
+  //cleaning (if we want zeros, we can remove cleaning)
+  clusters.forEach((ele) => {
+    const keys = Object.keys(ele);
+    for (key of keys) if (ele[key] === 0) delete ele[key];
+  });
+
+  return {
+    result: clusters,
+    allCount: clusters.reduce((sum, ele) => sum + (ele.All ?? 0), 0),
```

Comment:
`sumBy` is cleaner from lodash. We use the package anyway in other places

Reply:
I agree, I wanted to use typescript and npm packages inside the service worker, but unfortunately i could not use imports and exports within it,

`Uncaught SyntaxError: Cannot use import statement outside a module `;
`sw.js:4 Uncaught ReferenceError: exports is not defined` even when i was not exporting anything from service worker.